File: ReleaseNotes.API/ReleaseNotes.API/Services/Release/ReleaseService.cs
using ReleaseNotes.API.Data;
using ReleaseNotes.API.Services.Release.Models;
using Microsoft.EntityFrameworkCore;

namespace ReleaseNotes.API.Services.Release;

public class ReleaseService(DatabaseContext db) : IReleaseService
{
    public async Task<List<ReleaseResponseModel>> GetRecentReleases()
    {
        return await db.Releases.AsNoTracking()
            .Where(x => x.PublishedOn != null)
            .OrderByDescending(r => r.PublishedOn)
            .Take(5)
            .Select(x => new ReleaseResponseModel
            {
                Id = x.Id,
                Version = x.Version,
                ProjectName = x.Project!.Name,
                ProjectId = x.ProjectId,
            })
            .ToListAsync();
    }

    public async Task<bool> PublishRelease(Guid releaseId)
    {
        var release = await db.Releases.Include(x => x.Project).FirstOrDefaultAsync(x => x.Id == releaseId);

        if (release == null)
        {
            return false;
        }

        release.PublishedOn = DateTimeOffset.UtcNow;
        release.Project!.LatestActivity = DateTimeOffset.UtcNow;
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> UnpublishRelease(Guid releaseId)
    {
        var release = await db.Releases.Include(x => x.Project).FirstOrDefaultAsync(x => x.Id == releaseId);

        if (release == null)
        {
            return false;
        }

        release.PublishedOn = null;
        release.Project!.LatestActivity = DateTimeOffset.UtcNow;
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> CreateRelease(Guid projectId, CreateReleaseRequestModel createReleaseRequestModel)
    {
        var project = await db.Projects.FirstOrDefaultAsync(x => x.Id == projectId);
        if (project == null)
        {
            return false;
        }

        project.LatestActivity = DateTimeOffset.UtcNow;

        var release = new Data.Models.Release()
        {
            ProjectId = projectId,
            Version = createReleaseRequestModel.Version,
            CreatedOnUtc = DateTimeOffset.UtcNow,
        };

        db.Releases.Add(release);
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> EditRelease(Guid releaseId, EditReleaseRequestModel editReleaseRequestModel)
    {
        var release = await db.Releases.Include(x => x.Project).FirstOrDefaultAsync(x => x.Id == releaseId);
        if (release == null)
        {
            return false;
        }
        release.Project!.LatestActivity = DateTimeOffset.UtcNow;
        release.CreatedOnUtc = editReleaseRequestModel.CreatedOnUtc.ToUniversalTime();
        release.Version = editReleaseRequestModel.Version;
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> DeleteRelease(Guid releaseId)
    {
        var release = await db.Releases.Include(x => x.Project).Include(x => x.Notes).FirstOrDefaultAsync(x => x.Id == releaseId);
        if (release == null)
        {
            return false;
        }
        release.Project!.LatestActivity = DateTimeOffset.UtcNow;
        db.RemoveRange(release.Notes);
        db.Remove(release);
        await db.SaveChangesAsync();
        return true;
    }
}

File: ReleaseNotes.UI/src/services/release/release.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ReleaseResponseModel } from './models/release-response-model';
import { CreateReleaseRequestModel } from './models/create-release-request-model';
import { EditReleaseRequestModel } from './edit-release-request-model';

@Injectable({
  providedIn: 'root'
})
export class ReleaseService {

  private apiUrl = '/api/release';

  constructor(private http: HttpClient) { }

  getRecentReleases(): Observable<ReleaseResponseModel[]> {
    return this.http.get<ReleaseResponseModel[]>(`${this.apiUrl}/recent`);
  }

  publishRelease(releaseId: string) {
    return this.http.post(`${this.apiUrl}/${releaseId}/publish`, null);
  }

  unpublishRelease(releaseId: string) {
    return this.http.post(`${this.apiUrl}/${releaseId}/unpublish`, null);
  }

  createRelease(projectId: string, createReleaseRequestModel: CreateReleaseRequestModel) {
    return this.http.post(`${this.apiUrl}/${projectId}`, createReleaseRequestModel);
  }

  editRelease(releaseId: string, editReleaseRequestModel: EditReleaseRequestModel) {
    return this.http.put(`${this.apiUrl}/${releaseId}`, editReleaseRequestModel);
  }

  deleteRelease(releaseId: string) {
    return this.http.delete(`${this.apiUrl}/${releaseId}`);
  }
}
